### ISSUE TITLE
Memory error and matlab dir error with 0.12.0.rc1. 

### ISSUE THREAD

**Issue by Haakon-Ramsland** [nipy/nipype#1502]
Hi, I am real novice in this, and github. 

System: Ubuntu 14.04.4, old AMD system, Kernel 3.19.0.59.

I installed according to the guide on nipype, during the test run, i got the aniso2iso is missing error, so i updated by using the mastergit, and have currently 0.12.0.rc1 installed. 
I got the spm error, so i manually defined the matlab dir. in ipython. 

I get: 
Ran 17217 tests in 335.547s
FAILED (SKIP=15, errors=1, failures=3)

If i manually define the matlab dir.: 
MatlabCommand.set_default_paths('/usr/local/MATLAB/R2015b/toolbox/spm12'), i get these following errors: 
## FAIL: nipype.interfaces.tests.test_base.test_global_CommandLine_output(False,)

Traceback (most recent call last):
  File "/home/james/anaconda2/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
AssertionError: False is not true
# 
## FAIL: nipype.interfaces.tests.test_matlab.test_cmdline('/usr/local/MATLAB/R2015b/bin/glnxa64/MATLAB -nodesktop -nosplash -singleCompThread -r "fprintf(1,\'Executing code at %s:\n\',datestr(now));ver,try,addpath(\'/usr/local/MATLAB//usr/local/MATLAB/R2015b/bin/glnxa64/MATLAB\');whos,catch ME,fprintf(2,\'MATLAB code threw an exception:\n\');fprintf(2,\'%s\n\',ME.message);if length(ME.stack) ~= 0, fprintf(2,\'File:%s\nName:%s\nLine:%d\n\',ME.stack.file,ME.stack.name,ME.stack.line);, end;end;;exit"', '/usr/local/MATLAB/R2015b/bin/glnxa64/MATLAB -nodesktop -nosplash -singleCompThread -r "fprintf(1,\'Executing code at %s:\n\',datestr(now));ver,try,whos,catch ME,fprintf(2,\'MATLAB code threw an exception:\n\');fprintf(2,\'%s\n\',ME.message);if length(ME.stack) ~= 0, fprintf(2,\'File:%s\nName:%s\nLine:%d\n\',ME.stack.file,ME.stack.name,ME.stack.line);, end;end;;exit"')

Traceback (most recent call last):
  File "/home/james/anaconda2/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/james/anaconda2/lib/python2.7/site-packages/numpy/testing/utils.py", line 375, in assert_equal
    raise AssertionError(msg)
AssertionError: 
Items are not equal:
 ACTUAL: '/usr/local/MATLAB/R2015b/bin/glnxa64/MATLAB -nodesktop -nosplash -singleCompThread -r "fprintf(1,\'Executing code at %s:\n\',datestr(now));ver,try,addpath(\'/usr/local/MATLAB//usr/local/MATLAB/R2015b/bin/glnxa64/MATLAB\');whos,catch ME,fprintf(2,\'MATLAB code threw an exception:\n\');fprintf(2,\'%s\n\',ME.message);if length(ME.stack) ~= 0, fprintf(2,\'File:%s\nName:%s\nLine:%d\n\',ME.stack.file,ME.stack.name,ME.stack.line);, end;end;;exit"'
 DESIRED: '/usr/local/MATLAB/R2015b/bin/glnxa64/MATLAB -nodesktop -nosplash -singleCompThread -r "fprintf(1,\'Executing code at %s:\n\',datestr(now));ver,try,whos,catch ME,fprintf(2,\'MATLAB code threw an exception:\n\');fprintf(2,\'%s\n\',ME.message);if length(ME.stack) ~= 0, fprintf(2,\'File:%s\nName:%s\nLine:%d\n\',ME.stack.file,ME.stack.name,ME.stack.line);, end;end;;exit"'

I guess it's some error due to the defining of the matlab dir. 

---
## FAIL: Test runtime profiler correctly records workflow RAM/CPUs consumption

Traceback (most recent call last):
  File "/home/james/anaconda2/lib/python2.7/site-packages/nipype/interfaces/tests/test_runtime_profiler.py", line 436, in test_function_profiling
    self.assertLessEqual(runtime_gb_err, allowed_gb_err, msg=mem_err)
AssertionError: Input memory: 1.000000 is not within 0.250 GB of runtime memory: 1.255730
-------------------- >> begin captured logging << --------------------
workflow: INFO: ['check', 'execution', 'logging']
workflow: INFO: Running in parallel.
workflow: INFO: Executing: resource_node ID: 0
callback: DEBUG: {"num_threads": 2, "estimated_memory_gb": 1.0, "name": "resource_node", "start": "2016-06-08 22:47:57.463137", "id": "resource_node"}
workflow: INFO: [Job finished] jobname: resource_node jobid: 0
callback: DEBUG: {"num_threads": 2, "finish": "2016-06-08 19:48:05.612608", "name": "resource_node", "runtime_threads": 2, "estimated_memory_gb": 1.0, "runtime_memory_gb": 1.2557296752929688, "id": "resource_node"}
--------------------- >> end captured logging << ---------------------

What could cause the problem? Since no one else has reported the error, i assume it's somewhere in my system.  Ran memory test for some hours without any errors popping up. 

[crash-20160608-224757-james-resource_node.txt](https://github.com/nipy/nipype/files/305738/crash-20160608-224757-james-resource_node.txt)

Kind regards.
Haakon

**Comment by satra** [comment]
@Haakon-Ramsland - could you please post the error you get without setting anything?

from the terminal command line you can simply do: `nosetests --with-doctest nipype`

**Comment by Haakon-Ramsland** [comment]
Hi, thanks for the reply. 
Follwoing fail:

======================================================================
FAIL: Test runtime profiler correctly records workflow RAM/CPUs consumption
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/james/anaconda2/lib/python2.7/site-packages/nipype/interfaces/tests/test_runtime_profiler.py", line 436, in test_function_profiling
    self.assertLessEqual(runtime_gb_err, allowed_gb_err, msg=mem_err)
AssertionError: Input memory: 1.000000 is not within 0.250 GB of runtime memory: 1.339115
-------------------- >> begin captured logging << --------------------
workflow: INFO: ['check', 'execution', 'logging']
workflow: INFO: Running in parallel.
workflow: INFO: Executing: resource_node ID: 0
callback: DEBUG: {"num_threads": 2, "estimated_memory_gb": 1.0, "name": "resource_node", "start": "2016-06-09 18:14:10.156815", "id": "resource_node"}
workflow: INFO: [Job finished] jobname: resource_node jobid: 0
callback: DEBUG: {"num_threads": 2, "finish": "2016-06-09 18:14:19.489986", "name": "resource_node", "runtime_threads": 2, "estimated_memory_gb": 1.0, "runtime_memory_gb": 1.3391151428222656, "id": "resource_node"}
--------------------- >> end captured logging << ---------------------

----------------------------------------------------------------------
Ran 17207 tests in 366.747s

FAILED (SKIP=19, errors=2, failures=1)

[crash-20160609-181410-james-resource_node.txt](https://github.com/nipy/nipype/files/307402/crash-20160609-181410-james-resource_node.txt)

Any suggestion?

Kind regards 
Haakon.

**Comment by chrisfilo** [comment]
@dclark87 could you have a look?

**Comment by Haakon-Ramsland** [comment]
When i look through the error log, I also have one error that might be relevant to the failure..

ERROR: Test runtime profiler correctly records workflow RAM/CPUs consumption
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/james/anaconda2/lib/python2.7/site-packages/nipype/interfaces/tests/test_runtime_profiler.py", line 374, in test_cmdline_profiling
    start_str, finish_str = self._run_cmdline_workflow(num_gb, num_threads)
  File "/home/james/anaconda2/lib/python2.7/site-packages/nipype/interfaces/tests/test_runtime_profiler.py", line 267, in _run_cmdline_workflow
    wf.run(plugin='MultiProc', plugin_args=plugin_args)
  File "/home/james/anaconda2/lib/python2.7/site-packages/nipype/pipeline/engine/workflows.py", line 595, in run
    runner.run(execgraph, updatehash=updatehash, config=self.config)
  File "/home/james/anaconda2/lib/python2.7/site-packages/nipype/pipeline/plugins/base.py", line 272, in run
    report_nodes_not_run(notrun)
  File "/home/james/anaconda2/lib/python2.7/site-packages/nipype/pipeline/plugins/base.py", line 93, in report_nodes_not_run
    raise RuntimeError(('Workflow did not execute cleanly. '
RuntimeError: Workflow did not execute cleanly. Check log for details
-------------------- >> begin captured logging << --------------------
callback: DEBUG: {"num_threads": 2, "estimated_memory_gb": 1.0, "name": "resource_node", "start": "2016-06-09 20:39:40.729736", "id": "resource_node"}
callback: DEBUG: {"num_threads": 2, "finish": "2016-06-09 20:39:40.774892", "name": "resource_node", "estimated_memory_gb": 1.0, "error": true, "id": "resource_node"}
--------------------- >> end captured logging << ---------------------

Hope this helps somewhat... 
[crash-20160609-203940-james-resource_node.txt](https://github.com/nipy/nipype/files/307692/crash-20160609-203940-james-resource_node.txt)

**Comment by dclark87** [comment]
So the failure is happening because the unit test asserts that the `runtime_memory_gb` is within 0.25 GB of `estimated_memory_gb`, as seen here: https://github.com/nipy/nipype/blob/master/nipype/interfaces/tests/test_runtime_profiler.py#L383-395. However, your runtime is ~0.339 GB different from the 1.0GB estimated. This discrepancy seems to vary based on the size of the test suite being run. You can change this line: https://github.com/nipy/nipype/blob/master/nipype/interfaces/tests/test_runtime_profiler.py#L137 to be something like 0.5GB to see if that helps. I apologize for this error - I have been trying to figure out what causes the difference in estimated and runtime memories and it is a difficult problem to troubleshoot.

**Comment by Haakon-Ramsland** [comment]
Thanks for the swift reply, I got failed result, but not failure message, just similar error message. 

ERROR: Test runtime profiler correctly records workflow RAM/CPUs consumption
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/james/anaconda2/lib/python2.7/site-packages/nipype/interfaces/tests/test_runtime_profiler.py", line 374, in test_cmdline_profiling
    start_str, finish_str = self._run_cmdline_workflow(num_gb, num_threads)
  File "/home/james/anaconda2/lib/python2.7/site-packages/nipype/interfaces/tests/test_runtime_profiler.py", line 267, in _run_cmdline_workflow
    wf.run(plugin='MultiProc', plugin_args=plugin_args)
  File "/home/james/anaconda2/lib/python2.7/site-packages/nipype/pipeline/engine/workflows.py", line 595, in run
    runner.run(execgraph, updatehash=updatehash, config=self.config)
  File "/home/james/anaconda2/lib/python2.7/site-packages/nipype/pipeline/plugins/base.py", line 272, in run
    report_nodes_not_run(notrun)
  File "/home/james/anaconda2/lib/python2.7/site-packages/nipype/pipeline/plugins/base.py", line 93, in report_nodes_not_run
    raise RuntimeError(('Workflow did not execute cleanly. '
RuntimeError: Workflow did not execute cleanly. Check log for details
-------------------- >> begin captured logging << --------------------
callback: DEBUG: {"num_threads": 2, "estimated_memory_gb": 1.0, "name": "resource_node", "start": "2016-06-09 21:41:36.909264", "id": "resource_node"}
callback: DEBUG: {"num_threads": 2, "finish": "2016-06-09 21:41:36.970429", "name": "resource_node", "estimated_memory_gb": 1.0, "error": true, "id": "resource_node"}
--------------------- >> end captured logging << ---------------------

----------------------------------------------------------------------
Ran 17214 tests in 406.030s

FAILED (SKIP=19, errors=1)
Out[6]: <nose.result.TextTestResult run=17214 errors=1 failures=0>

I guess that nipype can still be used, even though i get an error message or do you think crashes will occure do to this error?

Kind regards.
Haakon

**Comment by Haakon-Ramsland** [comment]
Sorry mate. I forgot to include in the previous message. 

Here you are.

Kind regards
Haakon


[crash-20160609-214136-james-resource_node.txt](https://github.com/nipy/nipype/files/309391/crash-20160609-214136-james-resource_node.txt)

**Comment by dclark87** [comment]
I apologize Haakon, but I am having trouble displaying the crash message. Does 'nipype_display_crash path_to_crash_file.pklz' return anything (mine is returning an import error)? If yours returns a valid traceback - do you mind piping the result to a txt file and uploading that here (via 'nipype_display_crash path_to_crash_file.pklz >> path_to_txt_file.txt')

**Comment by CataRay** [comment]
I believe I have the same problem:

`======================================================================
FAIL: Test runtime profiler correctly records workflow RAM/CPUs consumption
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/cata/anaconda2/lib/python2.7/site-packages/nipype/interfaces/tests/test_runtime_profiler.py", line 436, in test_function_profiling
    self.assertLessEqual(runtime_gb_err, allowed_gb_err, msg=mem_err)
AssertionError: Input memory: 1.000000 is not within 0.250 GB of runtime memory: 1.334545
-------------------- >> begin captured logging << --------------------
workflow: INFO: ['check', 'execution', 'logging']
workflow: INFO: Running in parallel.
workflow: INFO: Executing: resource_node ID: 0
callback: DEBUG: {"num_threads": 2, "estimated_memory_gb": 1.0, "name": "resource_node", "start": "2016-06-15 00:15:23.232168", "id": "resource_node"}
workflow: INFO: [Job finished] jobname: resource_node jobid: 0
callback: DEBUG: {"num_threads": 2, "finish": "2016-06-15 00:15:28.997333", "name": "resource_node", "runtime_threads": 2, "estimated_memory_gb": 1.0, "runtime_memory_gb": 1.3345451354980469, "id": "resource_node"}
--------------------- >> end captured logging << ---------------------

----------------------------------------------------------------------
Ran 17214 tests in 684.783s

FAILED (SKIP=15, errors=2, failures=2)
`
is there any solution?
Thank you !

**Comment by Haakon-Ramsland** [comment]
[crash-20160615-204949-james-resource_node.txt](https://github.com/nipy/nipype/files/316969/crash-20160615-204949-james-resource_node.txt)

Sorry for the delay in replying. It's a busy period at work.

Kind regards 
Brut.

**Comment by satra** [comment]
@dclark87 -  nosetests is failing on my OSX (10.11.4) instance

```
======================================================================
FAIL: Test runtime profiler correctly records workflow RAM/CPUs consumption
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/software/nipy-repo/nipype/nipype/interfaces/tests/test_runtime_profiler.py", line 396, in test_cmdline_profiling
    self.assertEqual(expected_runtime_threads, runtime_threads, msg=threads_err)
AssertionError: Input threads: 2 is not equal to runtime threads: 3
-------------------- >> begin captured logging << --------------------
workflow: INFO: ['check', 'execution', 'logging']
workflow: INFO: Running in parallel.
workflow: INFO: Executing: resource_node ID: 0
callback: DEBUG: {"num_threads": 2, "estimated_memory_gb": 1.0, "name": "resource_node", "start": "2016-06-17 09:43:38.275220", "id": "resource_node"}
workflow: INFO: [Job finished] jobname: resource_node jobid: 0
callback: DEBUG: {"num_threads": 2, "finish": "2016-06-17 09:43:46.652089", "name": "resource_node", "runtime_threads": 3, "estimated_memory_gb": 1.0, "runtime_memory_gb": 1.0074234008789062, "id": "resource_node"}
--------------------- >> end captured logging << ---------------------

======================================================================
FAIL: Test runtime profiler correctly records workflow RAM/CPUs consumption
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/software/nipy-repo/nipype/nipype/interfaces/tests/test_runtime_profiler.py", line 437, in test_function_profiling
    self.assertEqual(expected_runtime_threads, runtime_threads, msg=threads_err)
AssertionError: Input threads: 2 is not equal to runtime threads: 3
-------------------- >> begin captured logging << --------------------
workflow: INFO: ['check', 'execution', 'logging']
workflow: INFO: Running in parallel.
workflow: INFO: Executing: resource_node ID: 0
callback: DEBUG: {"num_threads": 2, "estimated_memory_gb": 1.0, "name": "resource_node", "start": "2016-06-17 09:43:46.683776", "id": "resource_node"}
workflow: INFO: [Job finished] jobname: resource_node jobid: 0
callback: DEBUG: {"num_threads": 2, "finish": "2016-06-17 09:43:54.370508", "name": "resource_node", "runtime_threads": 3, "estimated_memory_gb": 1.0, "runtime_memory_gb": 1.0035667419433594, "id": "resource_node"}
--------------------- >> end captured logging << ---------------------
```

**Comment by dclark87** [comment]
I am addressing this by adding tolerance to number of threads here: https://github.com/FCP-INDI/nipype/commit/cff9393d026026de11823cb53b1a3915105a3dd6

**Comment by dclark87** [comment]
Hi @Haakon-Ramsland, according to your crash output - the use_resources script (found here: https://github.com/nipy/nipype/blob/master/nipype/interfaces/tests/use_resources) can't be found on your computer for some reason. This is part of the nipype source, so not sure why it's not there. I would double check to make sure you have the complete source code before running more unit tests.

**Comment by dclark87** [comment]
Just as an overall note @satra @chrisfilo - the runtime profiling code is an additional feature that still has some kinks - but serves more as a rough guide as to how many resources a nipype workflow uses - and the exact amount can vary from machine to machine. It is a work in progress.

**Comment by satra** [comment]
@dclark87 - in addition i'm getting a lot of these:

```
160618-01:14:49,721 interface INFO:
	 Could not get resources used by process. Error: (pid=10178)
```

was this tested on os x at all?

**Comment by satra** [comment]
@carolFrohlich - could you please take a look at this thread?

**Comment by dclark87** [comment]
This was not thoroughly tested on OSX. Is that the full exception statement? I would think the line: https://github.com/FCP-INDI/nipype/blob/resource_multiproc/nipype/interfaces/base.py#L1340-1341 would return more information than that.

**Comment by satra** [comment]
@dclark87 - i did a few more tests and it looks like the errors were really coming from not having psutil installed. i'll try to hack something to allow not generating that many errors if it doesn't detect psutil.

**Comment by dclark87** [comment]
Hm - I figured this: https://github.com/FCP-INDI/nipype/blob/resource_multiproc/nipype/interfaces/base.py#L1357-1363 or this: https://github.com/FCP-INDI/nipype/blob/resource_multiproc/nipype/interfaces/utility.py#L464-471 would have caught that. Is this the version/branch you are using?

**Comment by dclark87** [comment]
Actually - those same try blocks are in the nipy-master branch as well

**Comment by Haakon-Ramsland** [comment]
dclark87:
I used the git version, and the file was there, but changed to the pip version, now i get another error. 

File "/home/james/anaconda2/lib/python2.7/site-packages/nipype/interfaces/bru2nii.py", line 35, in nipype.interfaces.bru2nii.Bru2
Failed example:
    converter.cmdline  # doctest: +ELLIPSIS
Expected:
    'Bru2 -o .../nipype/nipype/testing/data/brukerdir brukerdir'
Got:
    'Bru2 -o /home/james/anaconda2/lib/python2.7/site-packages/nipype/testing/data/brukerdir brukerdir'



When i installed it on my laptop with ubuntu 16.04, it just hangs after 10 sek. Nothing happens, no error or crashfile. Perhaps the Python 3.2 doest play nice. 

Seems to be very nice software when it works,  but I have problems with at least the testing part of the program. 

Thanks for all the help. 

Kind regards
Haakon Ramsland Hol

**Comment by shoshber** [comment]
One of the above errors just failed a circle ci build of mine: https://circleci.com/gh/shoshber/nipype/182

So I looked into it--it seems to be the same problem as https://github.com/nipy/nipype/issues/1607, so I've made another pr for the fix: 

(I /should/ have caught this last week when I fixed the other sporadic multiproc test failure.)

I believe this constitutes a +1 for https://github.com/nipy/nipype/issues/1657

**Comment by mgxd** [comment]
@Haakon-Ramsland the `Bru2` error should be fixed in latest master. Feel free to open up this issue if you still have problems
